# Integrating Python, SQL, and Tableau

A good analyst must:
    
    1) be able to manage information from a database (SQL)
    
    2) be able to apply mathematical and statistical tools (Python)
    
    3) and present results in an intuitive way (Tableau)

# The Business Task

Predict "Absenteeism" at a company during working hours (i.e. whether or not an employee can be expected to be missing for a specific number of hours in a given workday)
- "Absenteeism" is defined as absence from work during normal working hours, resulting in temporary incapacity to execute regular working activity

# DATA PREPROCESSING

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

Dataset preprocessed already in other Juptyer notebook, see 'Preprocessing Dataset'

In [2]:
data_preprocessed = pd.read_csv('C:\\Users\\feder\\Downloads\\original (2).csv')
data_preprocessed.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0         0         0         0         1            7                1   
1         0         0         0         0            7                1   
2         0         0         0         1            7                2   
3         1         0         0         0            7                3   
4         0         0         0         1            7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          0         2     1                          4  
1               31          0         1     0                          0  
2               31          0         0     0                          2  
3               24          0         2     0                          4  
4               30          0         2     1                          2

# MACHINE LEARNING

In [3]:
data = data_preprocessed.copy()

Logistic regression is a type of classification, so we will be classifying people into two groups:
- Excessively absent 
- Not excessively absent

We will define excessviely absent people as those who are above the median absenteeism time in hours

In [4]:
data['Absenteeism Time in Hours'].median()

3.0

# Step 1: Define Targets

If a person has an absenteeism time in hours <= 3, they get a 0

If a person has an absenteeism time in hours >= 4, they get a 1

These 0s and 1s are called the targets in supervised machine learning

The task is to predict whether we will obtain a 0 or a 1 based on the independent variables

In [5]:
targets = np.where(data['Absenteeism Time in Hours'] > data['Absenteeism Time in Hours'].median(), 1, 0)

In [6]:
data['Excessive Absenteeism'] = targets
data.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0         0         0         0         1            7                1   
1         0         0         0         0            7                1   
2         0         0         0         1            7                2   
3         1         0         0         0            7                3   
4         0         0         0         1            7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  \
0               30          0         2     1                          4   
1               31          0         1     0                          0   
2               31          0         0     0                          2   
3               24          0         2     0                          4   
4               30          0         2     1                          2   

   Excessive Absenteeism  
0                      1  
1                      0  
2                      0  
3                      1  
4                      0

Drop the 'Absenteeism Time in Hours' variable

In [7]:
data = data.drop(['Absenteeism Time in Hours'], axis=1)
data.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0         0         0         0         1            7                1   
1         0         0         0         0            7                1   
2         0         0         0         1            7                2   
3         1         0         0         0            7                3   
4         0         0         0         1            7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Excessive Absenteeism  
0               30          0         2     1                      1  
1               31          0         1     0                      0  
2               31          0         0     0                      0  
3               24          0         2     0                      1  
4               30          0         2     1                      0

# Step 2: Select Features (Inputs)

In [8]:
#everything but the target, 'Excessive Absenteeism'
features_unscaled = data.iloc[:,:-1]
features_unscaled.head()

Reason_1  Reason_2  Reason_3  Reason_4  Month Value  Day of the Week  \
0         0         0         0         1            7                1   
1         0         0         0         0            7                1   
2         0         0         0         1            7                2   
3         1         0         0         0            7                3   
4         0         0         0         1            7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  
0               30          0         2     1  
1               31          0         1     0  
2               31          0         0     0  
3               24          0         2     0  
4               30          0         2     1

# Step 3: Standardize the Data

In [9]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

In [10]:
scaler.fit(features_unscaled)

StandardScaler(copy=True, with_mean=True, with_std=True)

# For new data:

new_data_raw = pd.read_csv('new_data.csv')

new_data_scaled = scaler.transform(new_data_raw)

In [11]:
features = scaler.transform(features_unscaled)

# Step 4: Split into Training and Testing

In [12]:
from sklearn.model_selection import train_test_split

In [13]:
train_test_split(features, targets)

[array([[ 1.73205081, -0.09298136, -0.31448545, ..., -0.44798003,
          2.67996851, -0.58968976],
        [-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
         -0.01928035,  0.26848661],
        [-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
         -0.91902997, -0.58968976],
        ...,
        [-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
          2.67996851, -0.58968976],
        [-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
         -0.01928035, -0.58968976],
        [ 1.73205081, -0.09298136, -0.31448545, ...,  2.23224237,
         -0.91902997, -0.58968976]]),
 array([[-0.57735027, -0.09298136, -0.31448545, ...,  2.23224237,
          0.88046927, -0.58968976],
        [-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
         -0.01928035,  1.12666297],
        [-0.57735027, -0.09298136, -0.31448545, ..., -0.44798003,
         -0.01928035,  0.26848661],
        ...,
        [-0.57735027, -0.09298136, -0.31448545, ..., -

4 arrays: 
    - array 1 is training features
    - array 2 is training targets
    - array 3 is testing features
    - array 4 is testing targets

In [14]:
x_train, x_test, y_train, y_test = train_test_split(features, targets, train_size=0.9, test_size=0.1, random_state=20)

In [15]:
print(x_train.shape, y_train.shape)

(630, 14) (630,)


In [16]:
print(x_test.shape, y_test.shape)

(70, 14) (70,)


# Step 5: Train the Model

In [17]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

In [18]:
reg = LogisticRegression()

In [19]:
reg.fit(x_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [20]:
reg.score(x_train, y_train)

0.7746031746031746

Based on the data, the model learned to classify 77% of the observations correctly

# Step 6: Summary Table

In [21]:
reg.intercept_

array([-0.26690023])

In [22]:
reg.coef_

array([[ 2.07182947,  0.33242448,  1.55567164,  1.35921467,  0.17442534,
        -0.18042265,  0.71943571, -0.03216552, -0.25824789,  0.02533943,
         0.31101977, -0.09890818,  0.40640687, -0.39381569]])

In [23]:
feature_names = features_unscaled.columns.values
feature_names

array(['Reason_1', 'Reason_2', 'Reason_3', 'Reason_4', 'Month Value',
       'Day of the Week', 'Transportation Expense', 'Distance to Work',
       'Age', 'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets'], dtype=object)

In [24]:
summary_table = pd.DataFrame(columns=['Feature name'], data = feature_names)
summary_table['Coefficient'] = np.transpose(reg.coef_)
summary_table.index = summary_table.index + 1
summary_table.loc[0] = ['Intercept', reg.intercept_[0]]
summary_table = summary_table.sort_index()
summary_table['Odds_ratio'] = np.exp(summary_table.Coefficient)
summary_table.sort_values('Odds_ratio', ascending=False)

Feature name  Coefficient  Odds_ratio
1                  Reason_1     2.071829    7.939335
3                  Reason_3     1.555672    4.738268
4                  Reason_4     1.359215    3.893135
7    Transportation Expense     0.719436    2.053274
13                 Children     0.406407    1.501413
2                  Reason_2     0.332424    1.394345
11          Body Mass Index     0.311020    1.364816
5               Month Value     0.174425    1.190562
10  Daily Work Load Average     0.025339    1.025663
8          Distance to Work    -0.032166    0.968346
12                Education    -0.098908    0.905826
6           Day of the Week    -0.180423    0.834917
9                       Age    -0.258248    0.772404
0                 Intercept    -0.266900    0.765749
14                     Pets    -0.393816    0.674478

# Interpretation

If a features coefficient is around 0 or its odds ratio is around 1, it is not particularly important

This is because:
- A weight of 0 implies that no matter the feature value, we will multiply it by 0 in the model

and

- For a unit change in the standardized feature, the odds increase by a multiple equal to the odds ratio (1 = no change) 
- ex. if odds are 5:1 and odds ratio is 2, odds increase to 10:1
- ex. if odds are 5:1 and odds ratio is 0.2, odds decrease to 1:1

# Step 7: Testing the Model

In [25]:
reg.score(x_test, y_test)

0.7285714285714285

Based on data it has never seen before, the model can correctly predict if a person will be excessively absent 72.86% of the time

# Probabilities 

In [ ]:
#Use sklearn.linear_model.LogisticRegression.predict_proba(x) to return the probability estimates for all possible outputs

In [26]:
predicted_proba = reg.predict_proba(x_test)
predicted_proba

array([[0.77662339, 0.22337661],
       [0.63696168, 0.36303832],
       [0.35385559, 0.64614441],
       [0.84208814, 0.15791186],
       [0.05326752, 0.94673248],
       [0.30418467, 0.69581533],
       [0.34361761, 0.65638239],
       [0.09908607, 0.90091393],
       [0.81156008, 0.18843992],
       [0.77201525, 0.22798475],
       [0.47103179, 0.52896821],
       [0.12421597, 0.87578403],
       [0.04988931, 0.95011069],
       [0.69427333, 0.30572667],
       [0.29179082, 0.70820918],
       [0.54897843, 0.45102157],
       [0.53563908, 0.46436092],
       [0.59263561, 0.40736439],
       [0.4396301 , 0.5603699 ],
       [0.02446578, 0.97553422],
       [0.69881053, 0.30118947],
       [0.82518595, 0.17481405],
       [0.37449995, 0.62550005],
       [0.43314371, 0.56685629],
       [0.21174038, 0.78825962],
       [0.7626906 , 0.2373094 ],
       [0.49577698, 0.50422302],
       [0.89229991, 0.10770009],
       [0.1437471 , 0.8562529 ],
       [0.8068895 , 0.1931105 ],
       [0.

Left column = probability of being 0

Right column = probability of being 1

In [27]:
predicted_proba[:,1]

array([0.22337661, 0.36303832, 0.64614441, 0.15791186, 0.94673248,
       0.69581533, 0.65638239, 0.90091393, 0.18843992, 0.22798475,
       0.52896821, 0.87578403, 0.95011069, 0.30572667, 0.70820918,
       0.45102157, 0.46436092, 0.40736439, 0.5603699 , 0.97553422,
       0.30118947, 0.17481405, 0.62550005, 0.56685629, 0.78825962,
       0.2373094 , 0.50422302, 0.10770009, 0.8562529 , 0.1931105 ,
       0.36509567, 0.71478834, 0.71184555, 0.55475159, 0.17481405,
       0.51922284, 0.1917816 , 0.83069965, 0.44578018, 0.5930028 ,
       0.22985037, 0.41468363, 0.19443827, 0.09533008, 0.78496521,
       0.64993419, 0.70312438, 0.24970836, 0.17042677, 0.15244161,
       0.60496372, 0.09752602, 0.68615158, 0.28763163, 0.87053298,
       0.50469798, 0.92948596, 0.28283108, 0.06734963, 0.06187335,
       0.76133113, 0.66425895, 0.33061569, 0.84321947, 0.17146205,
       0.24621399, 0.01179962, 0.2368076 , 0.79018472, 0.34886414])

These are the individual probabilities of being excessively absent

# SAVING THE MODEL

save the 'reg' variable into a file that can be loaded into a new notebook so others can use the ML algorithim

In [28]:
import pickle

In [29]:
#'model' refers to the file name
#'wb' means write bytes, when we open the file we will use 'rb' or read bytes
with open('model', 'wb') as file:
    pickle.dump(reg, file)

In [30]:
# pickle the scaler file
with open('scaler','wb') as file:
    pickle.dump(scaler, file)